Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_wfssrv.cpp

```diff
@@ -0,0 +1,266 @@
+// Copyright 2016 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Core/IPC_HLE/WII_IPC_HLE_Device_usb_wfssrv.h"
+
+#include <string>
+#include <vector>
+
+#include "Common/CommonTypes.h"
+#include "Common/FileUtil.h"
+#include "Common/Logging/Log.h"
+#include "Common/NandPaths.h"
+#include "Core/HW/DVDInterface.h"
+#include "Core/HW/Memmap.h"
+
+namespace WFS
+{
+std::string NativePath(const std::string& wfs_path)
+{
+  return File::GetUserPath(D_WFSROOT_IDX) + Common::EscapePath(wfs_path);
+}
+}
+
+CWII_IPC_HLE_Device_usb_wfssrv::CWII_IPC_HLE_Device_usb_wfssrv(u32 device_id,
+                                                               const std::string& device_name)
+    : IWII_IPC_HLE_Device(device_id, device_name)
+{
+  m_device_name = "msc01";
+}
+
+IPCCommandResult CWII_IPC_HLE_Device_usb_wfssrv::Open(u32 command_address, u32 mode)
+{
+  INFO_LOG(WII_IPC_HLE, "/dev/usb_wfssrv: Open");
+
+  // TODO(wfs): Proper GetTitleId() implementation which takes into account
+  // nandloader content too. Refactor with ES.
+  u64 title_id;
+  DVDInterface::GetVolume().GetTitleID(&title_id);
+
+  Memory::Write_U32(title_id >> 32, command_address + 4);
+  m_Active = true;
+  return GetDefaultReply();
+}
+
+IPCCommandResult CWII_IPC_HLE_Device_usb_wfssrv::Close(u32 command_address, bool force)
+{
+  INFO_LOG(WII_IPC_HLE, "/dev/usb_wfssrv: Close");
+  if (!force)
+    Memory::Write_U32(FS_SUCCESS, command_address + 4);
+  m_Active = false;
+  return GetDefaultReply();
+}
+
+IPCCommandResult CWII_IPC_HLE_Device_usb_wfssrv::IOCtl(u32 command_address)
+{
+  u32 command = Memory::Read_U32(command_address + 0xC);
+  u32 buffer_in = Memory::Read_U32(command_address + 0x10);
+  u32 buffer_in_size = Memory::Read_U32(command_address + 0x14);
+  u32 buffer_out = Memory::Read_U32(command_address + 0x18);
+  u32 buffer_out_size = Memory::Read_U32(command_address + 0x1C);
+
+  int return_error_code = FS_SUCCESS;
+
+  switch (command)
+  {
+  case IOCTL_WFS_INIT:
+    // TODO(wfs): Implement.
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_INIT");
+    break;
+
+  case IOCTL_WFS_DEVICE_INFO:
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_DEVICE_INFO");
+    Memory::Write_U64(16ull << 30, buffer_out);  // 16GB storage.
+    Memory::Write_U8(4, buffer_out + 8);
```

Comment:
I'm not sure yet, all I know is that this is what the PowerPC code expects as the return value. I'm guessing it's either the status or the type of the MSC device.